### PR TITLE
Improve LLVM compile and link configuration

### PR DESCRIPTION
**Pull request by mppf**
### Remove gather-clang-sysroot-arguments
For #18530

gather-clang-sysroot arguments gathers `-isysroot` and `-resource-dir` arguments from `clang` and these are necessary for a bundled `clang` compiler to function on Mac OS X. This PR changes this logic to run as part of `printchplenv` which reduces the probability that a bundled-LLVM build will stop functioning if the `clang` tools change. A reasonable alternative would be to set these when configuring the bundled LLVM build with cmake.

The `-isysroot` and `-resource-dir` arguments seem to no longer be necessary for Homebrew installs of LLVM 11, so the logic now focuses on the bundled configuration.
 
With these changes, we no longer need `gather-clang-sysroot-arguments`, `CLANG_SETTINGS`, or `configured-clang-sysroot-arguments`

### Adjust PrgEnv integration for LLVM backend
For https://github.com/Cray/chapel-private/issues/1817 and https://github.com/Cray/chapel-private/issues/2571
See PR #12202 for history in this area.

We have been running into multiple problems with the Chapel module on a Cray XC/EX. Issues include
 * the bundled clang using a linker incompatible with the libraries gathered from loaded modules
 * the bundled clang tries to use a GCC installation that is not present on the installed system.
 
To fix these errors, this PR takes the following steps:
 * stops setting the cmake option `-DGCC_INSTALL_PREFIX` since that effectively bakes the paths to that GCC installation into the built clang and causes problems if that path is not present (say if the module is installed somewhere without it)
 * instead, compute the GCC prefix in printchplenv and pass it to clang with `--gcc-toolchain` when it is unusual
 * instead of switching to PrgEnv-gnu only within `gather-cray-prgenv-arguments.bash`, don't switch PrgEnvs. The PR includes off-by-default support for running the whole compiler in PrgEnv-gnu if another PrgEnv is loaded. To access that, run with `CHPL_DO_PE_ENV_CHECK=1`. Either way, it's no longer the case that the gathered libraries are inconsistent with the linker used.

Also, this PR migrates the `gather-cray-prgenv-arguments.bash` logic to the Python `chplenv` / `printchplenv` scripts. This makes it more debuggable and fit in better with other configuration. Along with this, this PR changes strategy from using `cc -craype-verbose -###` to using `cc --cray-print-opts=cflags` and `cc --cray-print-opts=libs` which are more intended for getting the arguments. (Note that `gather-cray-prgenv-arguments.bash` is still used in portions of the runtime build. Addressing that is future work.)

Additionally, this PR takes some steps to follow up to #18485. That PR added a little table in the `chplenv` Python scripts to map from `craype-` CPU types to the names that clang uses. But, this table might be out of date. So, this PR:
 * adds code to query the LLVM infrastructure to see if the `CHPL_TARGET_CPU` value is a legal clang value. If it isn't, issue a warning and continue. (Clang normally halts compilation in this case).
 * adjusts the `printchplenv` scripts to allow `CHPL_TARGET_CPU` to be set to override the `craype-` value if the LLVM backend is being used.

Lastly, this PR removes some code in the compiler that was setting `-static` explicitly for CLE 6 systems. This does not seem to be necessary anymore in my testing.

### Default to CHPL_TARGET_CPU=native for local arm
Follow-up to PR #18485 for https://github.com/Cray/chapel-private/issues/2529.
Resolves https://github.com/Cray/chapel-private/issues/2529

Previously, `chplenv` scripts prevented `CHPL_TARGET_CPU=native` from working or being the default on ARM platforms as a workaround for a problem in `clang` with that configuration. This workaround should no longer be necessary since LLVM 7 added `-mcpu=native` support for ARM platforms.

- [x] verified that with LLVM backend and `CHPL_TARGET_CPU=native` inferred, we are passing `-mcpu=native` to clang with `--fast` on aarch64

### Add functions to version.h to return version_num info

This is just to avoid needing to recompile preFold.cpp when the compiler version number changes (usually with a different git revision).

### Testing

* Note, when building the compiler itself (i.e. `make`), LLVM libraries don't seem to link correctly when mixing PrgEnv-gnu and PrgEnv-cray (E.g. when object files are built with PrgEnv-gnu and then we try to link with PrgEnv-cray).

- [x] full local testing

Does `hello.chpl` compile in the default configuration (including LLVM backend) with:
- [x] SLES 12, system LLVM
- [x] SLES 12, bundled LLVM, system LLVM not in PATH etc
- [x] Ubuntu 21.04, system LLVM
- [x] Ubuntu 21.04, bundled LLVM
- [x] Mac OS X Mojave, system LLVM
- [x] Mac OS X Mojave, bundled LLVM
- [x] CLE 6, bundled LLVM, PrgEnv-gnu, without hugepages loaded
- [x] CLE 6, bundled LLVM, PrgEnv-gnu, with hugepages loaded (`module load craype-hugepages16M`)
- [x] CLE 6, bundled LLVM, make with PrgEnv-gnu, switch to PrgEnv-cray, with hugepages loaded. Works without `CHPL_DO_PE_ENV_CHECK`.
- [x] CLE 6, bundled LLVM, make with PrgEnv-gnu, switch to PrgEnv-intel, with hugepages loaded. Works without `CHPL_DO_PE_ENV_CHECK`.
- [x] CLE 7, bundled LLVM, PrgEnv-gnu, without hugepages loaded
- [x] CLE 7, bundled LLVM, PrgEnv-gnu, with hugepages loaded
- [x] CLE 7, bundled LLVM, make with PrgEnv-gnu, switch to PrgEnv-cray, with hugepages loaded. Works without `CHPL_DO_PE_ENV_CHECK`.
- [x] CLE 7, bundled LLVM, make with PrgEnv-gnu, switch to PrgEnv-intel, with hugepages loaded. Works without `CHPL_DO_PE_ENV_CHECK`.

Reviewed by @jhh67 and @ronawho - thanks!

### CODE REVIEW

**Comment by mppf** [comment]
@bradcray - could you try this out on Mojave? Thanks.

**Comment by bradcray** [comment]
No joy...  It looks like my nemesis is back:

```
$ chpl test/release/examples/hello.chpl 
ld: unknown option: -platform_version
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
error: Make Binary - Linking
```

And interestingly, it seems that the key missing flag `-mlinker-version=450` is present on a few of the internal commands, but not on the final link command:

```
<internal clang code generation> -mlinker-version=450 ...
 
<chapel clang driver invocation> -mlinker-version=450

...

# Make Binary - Linking
/usr/local/Cellar/llvm@11/11.1.0_2/bin/clang++  ...[no -mlinker-version=450 anywhere in here...]
```

Typing this last command manually with `-mlinker-version=450` added seems to work as expected.

**Comment by bradcray** [comment]
It does seem to work now (and I'm seeing the required flag), thanks Michael!

**Comment by mppf** [comment]
OK, now we've done all of the testing I am expecting to do here pre-merge. Now the PR just needs a review.

**Review by jhh67 — APPROVED** [review]
There is a lot going on here, but I didn't see any obvious problems.

File: compiler/main/driver.cpp [hunk 1]
```diff
@@ -1686,6 +1694,58 @@ static void validateSettings() {
   checkUnsupportedConfigs();
 }
 
+static void maybeRelaunchInPrgEnv(int argc, char* argv[]) {
+  // Relaunch in PrgEnv-gnu if:
+  //  * platform is cray-x* or cray-ex
+  //  * CHPL_TARGET_COMPILER is llvm
+  //  * PE_ENV does not indicate PrgEnv-gnu is loaded
```
Comment (ronawho) [inline compiler/main/driver.cpp:1701]: Can you add a top-level comment about this being off by default and opt-in with `CHPL_DO_PE_ENV_CHECK`? Looking quickly I assumed this was done by default, and only knew otherwise because you told me offline.

File: util/chplenv/chpl_llvm.py [hunk 2]
```diff
@@ -268,6 +272,224 @@ def llvm_enabled():
...
+    return (comp_args, link_args)
+
+@memoize
+def get_clang_compile_args():
+    (comp_args, link_args) = get_clang_compile_link_args()
+    return " ".join(comp_args)
+
+@memoize
+def get_clang_link_args():
+    (comp_args, link_args) = get_clang_compile_link_args()
```
Comment (ronawho) [inline util/chplenv/chpl_llvm.py:490]: Typically in python for unused args in a tuple `_` is used instead. So like:

```python
(_, link_args) = get_clang_compile_link_args()
```

File: util/config/gather-cray-prgenv-arguments.bash [hunk 3]
```diff
@@ -8,6 +8,8 @@
 # $4 = CHPL_COMM_SUBSTRATE
```
Comment (ronawho) [inline util/config/gather-cray-prgenv-arguments.bash:8]: Is this file still needed?

Comment (mppf) [reply util/config/gather-cray-prgenv-arguments.bash:8]: Yes, it's used in the runtime build, for now at least. Future work to remove it.
